routes/runtime/github.rb

```diff
@@ -0,0 +1,115 @@
+    end
+
+    r.get "cache" do
+      fail CloverError.new(204, "NotFound", "No cache entry") if repository.access_key.nil?
+
+      keys, version = r.params["keys"]&.split(","), r.params["version"]
+      fail CloverError.new(400, "InvalidRequest", "Wrong parameters") if keys.nil? || keys.empty? || version.nil?
+
+      scopes = [runner.workflow_job&.dig("head_branch"), repository.default_branch].compact
+      entry = repository.cache_entries_dataset
+        .exclude(committed_at: nil)
+        .where(key: keys, version: version, scope: scopes).all
+        .min_by { keys.index(_1.key) + (scopes.index(_1.scope) * keys.size) }
+
+      fail CloverError.new(204, "NotFound", "No cache entry") if entry.nil?
+
+      entry.update(last_accessed_at: Time.now, last_accessed_by: runner.id)
+      signed_url, _ = repository.url_presigner.presigned_request(:get_object, bucket: repository.bucket_name, key: entry.blob_key)
+
+      {
+        scope: entry.scope,
+        cacheKey: entry.key,
+        cacheVersion: entry.version,
+        creationTime: entry.created_at,
+        archiveLocation: signed_url
+      }
+    end
+
+    r.on "caches" do
+      r.get true do
+        key = r.params["key"]
+        fail CloverError.new(204, "NotFound", "No cache entry") if key.nil?
+
+        scopes = [runner.workflow_job&.dig("head_branch"), repository.default_branch].compact
+        entries = repository.cache_entries_dataset
+          .exclude(committed_at: nil)
+          .where(key: key, scope: scopes).all
+
+        {
+          totalCount: entries.count,
+          artifactCaches: entries.map do
+            {
+              scope: _1.scope,
+              cacheKey: _1.key,
+              cacheVersion: _1.version,
+              creationTime: _1.created_at
+            }
+          end
+        }
+      end
+
+      # YYYY: Create bucket if not exists
+
+      r.post true do
+        key = r.params["key"]
+        version = r.params["version"]
+        size = Integer(r.params["cacheSize"])
+        fail CloverError.new(400, "InvalidRequest", "Wrong parameters") if key.nil? || version.nil? || size.nil?
+
+        unless (scope = runner.workflow_job&.dig("head_branch"))
+          # YYYY: If the webhook not delivered yet, we can try to get the branch from the API
+          Clog.emit("The runner does not have a workflow job") { {no_workflow_job: {ubid: runner.ubid, repository_ubid: repository.ubid}} }
+          fail CloverError.new(400, "InvalidRequest", "No workflow job data available")
+        end
+
+        entry = GithubCacheEntry.create_with_id(repository_id: runner.repository.id, key: key, version: version, size: size, scope: scope, created_by: runner.id)
+
+        upload_id = repository.blob_storage_client.create_multipart_upload(bucket: repository.bucket_name, key: entry.blob_key).upload_id
+        entry.update(upload_id: upload_id)
+
+        max_chunk_size = 32 * 1024 * 1024 # 32MB
```

Comment:
I left a comment regarding this in other repository as well. Should we use a static chink size or somehow calculate a dynamic chunk size?

Reply:
The GitHub cache client includes a configuration value for setting the maximum chunk size, which defaults to 32MB. It remains unchanged in most customer workflows. Given the varying multipart upload workflows of Azure Blob Storage and S3 compatible storages, the server also needs to be aware of this maximum chunk size. 

I believe there's no need for a more complex chunk size calculation as it's also fixed for the default GitHub cache. However, I've added the maximum chunk size value to reserveCache response, enabling the backend to determine it. This logic can be enhanced in the future without causing any disruption to customers.

Reply:
In our case, is the concurrency defined by the size of the cache entry divided by 32MB? How many concurrent upload connections would we get for the maximum allowed cache entry size?

Reply:
The cache client begins uploading all chunks simultaneously.

There was no mention of a maximum concurrency limit in the docs. The maximum number of upload parts allowed is 10,000. However, given our maximum cache size of 10GB/32MB, which equals 312 chunks, we can impose a concurrency limit on the client if necessary.

https://developers.cloudflare.com/r2/reference/limits/#account-plan-limits